CsprSdkStandardTestsNet/Test/Steps/SpeculativeExecutionSteps.cs

```diff
@@ -0,0 +1,344 @@
+        WriteLine("the speculative_exec execution_result contains a valid {0} transform with a value of {1}", type, value);
+
+        var speculativeDeployData =
+            _contextMap.Get<RpcResponse<SpeculativeExecutionResult>>(StepConstants.DEPLOY_RESULT).Parse();
+
+        var transform = speculativeDeployData.ExecutionResult.Effect.Transforms.Last();
+        
+        Assert.That(transform.Type, Is.EqualTo(TransformType.AddUInt512));
+        Assert.That(transform.Value, Is.EqualTo(BigInteger.Parse(value.ToString())));
+        
+    }
+    
+    [Then(@"the speculative_exec execution_result contains a valid balance transform")]
+    public void ThenTheSpeculativeExecExecutionResultContainsAValidBalanceTransform() {
+        WriteLine("he speculative_exec execution_result contains a valid balance transform");
+        
+        var transforms = GetFaucetBalanceTransforms();
+        Assert.That(transforms.First().Key.KeyIdentifier, Is.EqualTo(KeyIdentifier.Balance));
+
+    }
+
+    private string GetPrivateKey(string user) {
+        return (user.Equals("faucet"))
+            ? AssetUtils.GetFaucetAsset(1, "secret_key.pem")
+            : AssetUtils.GetUserKeyAsset(1, int.Parse(user[^1..]), "secret_key.pem");
+    }
+
+    private async Task<GetAccountInfoResult> GetAccountInfo(string account) {
+
+        var speculativeDeployData =
+            _contextMap.Get<RpcResponse<SpeculativeExecutionResult>>(StepConstants.DEPLOY_RESULT).Parse();
+        
+        var stateAccount = await GetCasperService().GetAccountInfo(
+            KeyPair.FromPem(GetPrivateKey(account)).PublicKey, speculativeDeployData.BlockHash);
+
+        return stateAccount.Parse();
+
+    }
+
+    private DeployInfo GetDeployTransform() {
+        
+        var deploy = _contextMap.Get<Deploy>(StepConstants.DEPLOY);
+        var speculativeDeployData =
+            _contextMap.Get<RpcResponse<SpeculativeExecutionResult>>(StepConstants.DEPLOY_RESULT).Parse();
+        var key = "deploy-" + deploy.Hash.ToUpper();
+        
+        var transform =
+            speculativeDeployData.ExecutionResult.Effect.Transforms.Find(t => t.Key.ToString().ToUpper().Equals(key.ToUpper()));   
+        
+        Assert.That(transform, Is.Not.Null);
+        Assert.That(transform.Key.ToString().ToUpper(), Is.EqualTo(key.ToUpper()));
+            
+        return (DeployInfo)transform.Value;
+        
+    }
+
+    private List<Transform> GetFaucetBalanceTransforms() {
+        var mainPurse = GetAccountInfo("faucet").Result.Account.MainPurse.ToString().Split("-")[1];
+        var speculativeDeployData =
+            _contextMap.Get<RpcResponse<SpeculativeExecutionResult>>(StepConstants.DEPLOY_RESULT).Parse();
+        
+        var transforms = speculativeDeployData.ExecutionResult.Effect.Transforms.Where
+            (t => t.Key.ToString().ToUpper().Equals(("balance-" + mainPurse.ToUpper()).ToUpper()));
```

Comment:
Duplicate ToUpper: mainPurse.ToUpper()).ToUpper() why not use ToLower? as ""balance-"  is lowercase

Reply:
done